src/main/scala/outwatch/package.scala

```diff
@@ -0,0 +1,54 @@
+import rxscalajs.Observable
+
+package object outwatch {
+  type Pipe[-I, +O] = Observable[O] with Sink[I]
+  type Handler[T] = Pipe[T, T]
+
+  implicit class PipeOps[-I, +O](val self: Pipe[I, O]) extends AnyVal {
+
+    def mapSink[I2](f: I2 => I): Pipe[I2, O] = Pipe(self.redirectMap(f), self)
+
+    def mapSource[O2](f: O => O2): Pipe[I, O2] = Pipe(self, self.map(f))
+
+    def mapPipe[I2, O2](f: I2 => I)(g: O => O2): Pipe[I2, O2] = Pipe(self.redirectMap(f), self.map(g))
+
+
+    def collectSink[I2](f: PartialFunction[I2, I]): Pipe[I2, O] = Pipe(self.redirect(_.collect(f)), self)
+
+    def collectSource[O2](f: PartialFunction[O, O2]): Pipe[I, O2] = Pipe(self, self.collect(f))
+
+    def collectPipe[I2, O2](f: PartialFunction[I2, I])(g: PartialFunction[O, O2]): Pipe[I2, O2] = Pipe(
+      self.redirect(_.collect(f)), self.collect(g)
+    )
+
+
+    def filterSource(f: O => Boolean): Pipe[I, O] = Pipe(self, self.filter(f))
+
+
+    def transformSink[I2](f: Observable[I2] => Observable[I]): Pipe[I2, O] = Pipe(self.redirect(f), self)
+
+    def transformSource[O2](f: Observable[O] => Observable[O2]): Pipe[I, O2] = Pipe(self, f(self))
+
+    def transformPipe[I2, O2](f: Observable[I2] => Observable[I])(g: Observable[O] => Observable[O2]): Pipe[I2, O2] =
+      Pipe(self.redirect(f), g(self))
+  }
+
+  // This is not in PipeOps, because I is contravariant
+  implicit class FilterSink[I, +O](val self: Pipe[I, O]) extends AnyVal {
+    def filterSink(f: I => Boolean): Pipe[I, O] = Pipe(self.redirect(_.filter(f)), self)
+  }
+
+  implicit class HandlerOps[T](val self: Handler[T]) extends AnyVal {
+
+    def imap[S](read: T => S)(write: S => T): Handler[S] = self.mapPipe(write)(read)
```

Comment:
We should have the sink/source transformation order consistent between `Pipe` and `Handler`. Either transform the sink first everywhere, or the source first everywhere. Intuitively for me was to have the sink first, because something sent into the sink gets emitted by the observable afterwards.

Reply:
Good catch! Yes this should be consistent. My motivation for this order was that you can easily rewrite a `handler.map(f)` to `handler.imap(f)(g)` by just appending the second function.

Reply:
Interestingly

`mapPipe` is consistent with `Profunctor.dimap`:
https://github.com/typelevel/cats/blob/0997f43ab68aac1ea83ad0dbc0819d2ad3e5ef22/core/src/main/scala/cats/arrow/Profunctor.scala#L29

And `imap` with `Invariant.imap`:
https://github.com/typelevel/cats/blob/8d7b39407b282781c06ca580479b6cc125bcb312/core/src/main/scala/cats/Invariant.scala#L9

:sweat_smile:

Reply:
That's interesting. In that case, let's leave them as they are.